mrp_bom_nested_bom_generator/models/mrp_nested_bom.py

```diff
@@ -0,0 +1,196 @@
+        store=True,
+    )
+
+    product_tmpl_id = fields.Many2one(
+        comodel_name="product.template",
+        string="Product",
+        domain="[('id', '!=', bom_product_tmpl_id)]",
+    )
+    product_qty = fields.Float(
+        string="Qty",
+        default=1,
+        required=True,
+    )
+    uom_id = fields.Many2one(
+        "uom.uom",
+        string="Unit of Measure",
+        related="product_tmpl_id.uom_id",
+        readonly=True,
+    )
+
+    attribute_ids = fields.Many2many(
+        "product.attribute",
+        "attribute_ids_ref",
+        string="Attributes",
+    )
+
+    @api.onchange("product_tmpl_id")
+    def _onchange_product_tmpl_id(self) -> None:
+        """
+        Change attribute field at changing product template field
+        :return None
+        """
+        self.attribute_ids = self.product_tmpl_id.attribute_line_ids.mapped(
+            "attribute_id"
+        )
+
+    @api.depends("bom_id.nested_bom_ids", "bom_product_tmpl_id.attribute_line_ids")
+    def _compute_attribute_aggregated_ids(self) -> None:
+        """
+        Aggregate own attributes with parent product template ones
+        :return None
+        """
+        for rec in self:
+            default_attributes = rec.product_tmpl_id.attribute_line_ids.mapped(
+                "attribute_id"
+            )
+            parent_attributes = rec.bom_product_tmpl_id.attribute_line_ids.mapped(
+                "attribute_id"
+            )
+            rec.attribute_aggregated_ids = default_attributes | parent_attributes
+
+    @api.model
+    def create_product(self, bom) -> int:
+        """
+        Create product by parent product
+        :param bom: mrp.bom record
+        :return: product template record id or False
+        :rtype int or bool
+        """
+        if not bom:
+            return False
+        nested_bom_count: int = self.search_count(
+            [("bom_product_tmpl_id", "=", bom.product_tmpl_id.id)]
+        )
+        return (
+            self.env["product.template"]
+            .create(
+                {
+                    "name": "{} #{}".format(
+                        bom.product_tmpl_id.name, nested_bom_count + 1
+                    )
+                }
+            )
+            .id
+        )
+
+    @api.model
+    def create(self, vals: Dict[str, Any]):
+        product_tmpl_id = vals.get("product_tmpl_id", False)
+        product_qty = vals.get("product_qty", False)
+        bom_id = vals.get("bom_id", False)
+        bom = self.env["mrp.bom"].search([("id", "=", bom_id)], limit=1)
+        if not bom:
+            raise models.MissingError(_("Nested BOM: Mrp BOM record not found!"))
+        # Create product template if product_tmpl_id not set
+        if not product_tmpl_id:
+            pt_id = self.create_product(bom)
+            if pt_id:
+                vals.update(product_tmpl_id=pt_id)
+
+        # Raising Exception if product qty less or equal zero
+        if product_qty <= 0.0:
```

Comment:
You can use an SQL constraint for checking this

Reply:
Resolved